Add initial PCB design

nbelzer opened the pull request:
This PR includes the KiCad project for the PCB design I had manufactured (fb452062ea3a40b8f8a60e07f3715b3deedb0bc6).  

![CleanShot 2025-04-13 at 13 42 30@2x](https://github.com/user-attachments/assets/87ad69e1-4b74-471b-a4dd-293e1879c954)

I have started to make some improvements on top of the original design that I plan to include in my next order.  Any suggestions or feedback is welcome.

Changes so far:
- Removed the pull-ups as they should operate on a 3.3v line (which the board doesn't have).  Instead I'm opting to manually solder two resistors from the 3.3v pin on the ESP32 to the data lines.
- Move the connectors down a bit, there is very little tolerance between the stepper connector and the physical slots that hold the PCB, this should give some more clearance.
- Add a notch to the board design, I believe this will help route the stepper motor cable.
- I attempted to use a DIP switch for the address selection but found it to take up more space than the current solution.  Open to suggestions here.

nbelzer committed:
Add manufactured PCB design

This commit includes the KiCad project for the PCB that I had
manufactured.

nbelzer committed:
Remove data line pull-ups

The SCL and SDL pull-ups need to happen on 3.3v, yet in the current
design they are attached to the 5v line.  Since they are only necessary
once per circuit (8 boards + esp32) I will leave them out of the board
design.

nbelzer committed:
Make space for stepper connector

Moves all components down to increase the space for the stepper motor
JST conector.

nbelzer committed:
Re-route U1 bypass capacitor

nbelzer committed:
Add notch to board design

This should help route the stepper and hall effect cables.

nbelzer committed:
Add gitignore for KiCad projects

nbelzer committed:
Update version number

nbelzer commented:
> Any thoughts to just adding 3-pin headers to A0,1 and 2 so jumpers can be used? I'm not sure if that's any less space than the DIPs

That is what the current design looks like.  Each address has 3-pins and a jumper is used to set the pin to high or low.

nbelzer commented:
As I was assembling my displays with the produced batch (version in fb452062ea3a40b8f8a60e07f3715b3deedb0bc6) I realized that communication between the modules is happening with a lot of errors.  After debugging and some manual soldering I believe I have traced this back to the pull-up logic. 

- When soldering the pads for the pull-up at 5v I seem to introduce a lot of noise as whenever a second board (without the pull-up pads soldered) is connected everything stops working. I believe this is related to the internal pull-ups (which I assumed were not present when designing the initial version), not sure if those can be controlled in software.

- The ESP32 that we're using in this project already has internal pull-ups for the SCL and SDA lines (I measured it at 7kΩ) going to 3.3v. In my initial design I am powering the PCF8575 at 5v which means the line will never reach `high as we'd need (0.7 x 5v =) 3.5v to register as high.

- I scratched through the 5v connections to the PCF8575 on a few of the prototype boards and instead hand soldered a 3.3v line from the ESP32 to them.  Initially this didn't work until I realized that the address pins also need to be set from the 3.3v line.  I was able to make this work with 2 boards.

Based on this I believe a small re-design is necessary to power the PCF8575 at 3.3v.  This would require me to step down the voltage on each board (or add an additional wire to the connectors between boards). My preference goes to stepping down the voltage.

nbelzer committed:
Add 3.3v line for PCF8575

nbelzer committed:
Use two pin address connectors with pulldown resistor

nbelzer committed:
Replace linear regulator with AP2112K

This regulator has a smaller footprint and seems to be better
optimised for our use case (more efficient, lower idle power draw,
less heat).

nbelzer committed:
Accomodate linear regulator and new address pins on PCB

nbelzer commented:
I've pushed an update to this branch with the 3.3v rails, wider tracks, and a two-pin per address solution using pulldown resistors.

![CleanShot 2025-05-07 at 17 02 43@2x](https://github.com/user-attachments/assets/106d3391-9f8f-4644-967b-a62c344d26aa)

nbelzer committed:
Add missing part numbers

nbelzer committed:
Minor pcb changes

- Updates the position of the entire board down,
- Adds an edge cut section for the WSH130 wiring.
- Moves the ULN2003 up a bit to clean up the connections.
- Updates silkscreen labels.

nbelzer committed:
Add missing part number for address pins

nbelzer commented:
I've ordered a second batch of boards with this design (https://github.com/jhoff/Split-Flap-Display/pull/22/commits/301094fd006527d082b3781ad475421dd8043410), will update this PR with my findings once they arrive.

nbelzer committed:
Correct AP2112K capacitors

nbelzer committed:
Add pull-ups between PCF8575 and ULN2003

This is necessary since the PCF8575 can only pull the output to 0, not
bring it up to 3.3v, so a pull-up is necessary to communicate with the
ULN2003 (which only responds to values 2.5v or above).

nbelzer committed:
Use 2x3 connector for address pins

Saves on board space and component count

nbelzer committed:
Rewire PCB

nbelzer commented:
![v2 design with hand soldered fixes](https://github.com/user-attachments/assets/9aed622e-4349-42b3-a491-6f1b1db4379d)

I recently received the v2 boards and found two flaws that I was able to correct with some manual soldering:
- The 5v -> 3.3v step was incorrectly routed, this is now fixed
- The connection between the PCF8575 and ULN2003 requires pull-ups as at 3.3v the signal between them only reaches 1v (whereas on 5v the signal reaches 1.5v, which is out of spec but worked on the previous iterations). This has also been addressed.

![v3 design](https://github.com/user-attachments/assets/a749b230-e756-48e6-999f-4546228994f2)

LIVESKILLER commented:
very good

LIVESKILLER commented:
![上传IMG_8984.jpeg...]()
I think 0805 capacitors and resistors can be used on this PCB, because the board has enough area and can fully accommodate them. The 0402 components are too small.

nbelzer commented:
My display is now working with the v2 boards and [the manual fix highlighted before](https://github.com/jhoff/Split-Flap-Display/pull/22#issuecomment-2933626984).

| Sideview | Assembly | Result |
| --- | --- | --- |
| ![sideview](https://github.com/user-attachments/assets/1a2e73c5-64b4-4f08-b180-428166caf231) | ![assembly](https://github.com/user-attachments/assets/5990dc6d-5fb2-4e1c-8cbe-c7ad54984843) | ![splitflap](https://github.com/user-attachments/assets/cf03033f-c800-4e71-b260-c07880b84c3c) |

I did notice a power issue with more than 6 displays connected.  The whole array works when booted from an unpowered state, but would be inconsistent (only the first few displays would operate) when previously powered in the last few minutes.  I believe this might be related to the capacitors that are on each board, potentially drawing too much power.  I noticed that the display is stable with 6 units connected, so I scaled down to that for now.

The connection between the units works but the JST cables I used are a bit fragile. Since the cables need to be quite short (so that they fit in the display itself) it makes it harder to assemble and disassemble the display. I have been looking into pogo-pins as a potential solution to remove any cables from between the modules.  Maybe even create some sort of PCB with exposed tracks (similar to the [DIYson](https://github.com/stevenbennett/Open-Task-Light/tree/main) [project](https://www.youtube.com/watch?v=ZJvm6Ro7Wek&t=596s)) that could be part of some sort of base for all the modules.

pearce1340 commented:
This looks fantastic - is version 3 working as expected? I'm keen to get my own manufactured using your design, did you use PCBWay? Any advice on specs for ordering them?

Thanks and great work!

nbelzer committed:
Eliminate 3.3v line from PCB

Based on the findings by FoxManJ the 3.3v line is not required if we
use a 4.7k Ohm pull-up to 5v for the SCL and SDA lines coming from the
ESP32.

By using this pull-up to 5V we eliminate the need for the PCF8575 to
run at 3.3v to understand the signals from the ESP32, which was the
reason I originally implemented the 3.3v line on each board.

nbelzer committed:
Use pinsocket and headers for board to board connector

Eliminates the JST cables between boards.

nbelzer committed:
Update pin and header parts

nbelzer committed:
Add optional data lines pull-up

Required as the ESP32 by default does not pull up the data lines to
5v.  Should only be activated once per circuit (display).

nbelzer committed:
Rename pull-up jumper

nbelzer commented:
I just finished up v4 and ordered several for myself. Changes include:
- Removed the 3.3v step-down in favor of 5v pull-ups on the SDA and SCL lines.
  - The required pull-ups are included on each board but should only be used once per circuit. 
- Increase the cut-out on the right for the cables leading to the magnetic sensor.
- Use pin sockets and headers for connecting the boards together, this is much easier than using JST cables which in my experience were finicky, easy to break, and difficult to store in the small amount of space we have. 
  - I believe this also significantly reduces the resistance on the power and data lines.  This seems to be the reason that my previous build could only power 6 out of the 8 units.

<img width="2623" height="2220" alt="CleanShot 2025-11-16 at 12 35 30@2x" src="https://github.com/user-attachments/assets/280be82c-90db-4ab9-92c7-cdadac712186" />

If the ordered boards work as expected, I'll add a README with instructions on how to order these boards for yourself.

bpejac commented:
Hello @nbelzer ,
How was the latest batch of PCB? Did you get around to test it?

I would love to get some PCBs made from your specs, but my electronics and kicad skills are very low, so I am not sure I could debug any issues myself.

nbelzer commented:
> Hello @nbelzer , How was the latest batch of PCB? Did you get around to test it?
> 
> I would love to get some PCBs made from your specs, but my electronics and kicad skills are very low, so I am not sure I could debug any issues myself.

@brunottonurb I received the batch but only got around to testing them this weekend. I ran into an issue powering more than 2 boards so hold of on ordering any for now.  If its an issue with the board I'll update the designs once I figure out what is wrong.

mrrivest commented:
Any update? I'd love to order a few!

bpejac commented:
Me too! @nbelzer Can I support you by paying for the next order of PCBs?

nbelzer commented:
@mrrivest, @brunottonurb thanks for the support, I still need to diagnose the issue with the current PCB design, I'll try to make some time for it this week and let you know what I find.

nbelzer committed:
Fix floating pull-ups

The optional pull-ups added in v4 added an unintentional floating node
that would tie SCL and SDA together whenever the pull-up pad was
unsoldered.  Causing the I2C bus to fail (surprisingly only after
connecting 3 modules).

nbelzer commented:
Finally had some time today and I was able to diagnose the issue with the v4 boards.  I accidentally created a floating node that would tie SCL and SDA together if the optional pull-up pad was unbridged.  I was able to solve this for the PCBs I already ordered by scraping the existing 2k2 resistors from 7 out of the 8 boards, leaving only the ones on the board where the pull-up is active.

I am now able to connect 8 modules together with no issue.  I'll update the PCB size (as they are currently a bit too small) and add a README soon with instructions on how to order these yourself.

Wiktor43 commented:
Hey nbelzer! 😄

Do you have an estimated time when the updated PCBs will be available to order and the files released?
I’ve been checking for a while  really looking forward to it!

Thanks!

nbelzer committed:
Remove pull-ups entirely

For now manually soldering pull-ups to the ESP32 remain an easier solution.

nbelzer committed:
Extend PCB to 37mm

nbelzer committed:
Bump printed version number

nbelzer commented:
<img width="2137" height="1742" alt="CleanShot 2026-03-01 at 19 46 46@2x" src="https://github.com/user-attachments/assets/56f7aeec-4eb4-409b-9307-a00b79a3daf7" />

I just pushed the v5 version to this branch. While I haven't ordered these myself (I still have some stock of the v4) I am confident they work. I've applied the following changes:
- Removed the SDA/SCL pull-ups using the solder pad (deferring to manually soldering these to the ESP32 for now)
- Extended the board width from 35 to 37mm. Each PCB board now matches the width of a single display unit.

I will soon wrap up this PR by adding a README with instructions on how to order these for yourself.

I do have some more ideas on how to improve the boards beyond this point and will continue to work on it when I find the time.

justinbadal commented:
@nbelzer thank you for your hard work. Is this ready to go?

DrewFerg11 commented:
No intention to step on @nbelzer's toes (I've been eagerly following the PR as well!), but I did [push some files/instructions for ordering custom pcbs](https://github.com/jhoff/Split-Flap-Display/pull/42). This pcb is different (although fairly similar to Nick's v5 from what I can see) and was designed by dowjames in the discord. 

I've ordered these pcbs two different times and have used them for months with no obvious issues.
